gobblin-runtime/src/main/java/gobblin/runtime/Fork.java

```diff
@@ -319,12 +317,12 @@ public String toString() {
     return "Fork: TaskId = \"" + this.taskId + "\" Index: \"" + this.index + "\" State: \"" + this.forkState + "\"";
   }
 
+  /**
+   * Tell this fork that the parent task is done. This is a second chance call if the parent
+   * task failed and didn't do so through the normal way of calling markParentTaskDone().
+   */ 
   @Override
   public void close() throws IOException {
-    // Tell this fork that the parent task is done. This is a second chance call if the parent
-    // task failed and didn't do so through the normal way of calling markParentTaskDone().
-    this.parentTaskDone = true;
```

Comment:
Yes, I see your point, see my comment below.

I think the problem here is that the original behavior of the `Fork` class allows a the `close` method to be invoked before `markParentTaskDone` is invoke. The original `close` implementation will ensure the `while (true)` loop in `Fork.processRecords` exits properly, even if `markParentTaskDone` has not been invoked. I think we need to preserve this behavior.